cobc/replace.c

```diff
@@ -0,0 +1,313 @@
+/*
+   Copyright (C) 2003-2022 Free Software Foundation, Inc.
+   Written by Fabrice Le Fessant
+
+   This file is part of GnuCOBOL.
+
+   The GnuCOBOL compiler is free software: you can redistribute it
+   and/or modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation, either version 3 of the
+   License, or (at your option) any later version.
+
+   GnuCOBOL is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GnuCOBOL.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "tarstamp.h"
+#include "config.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdarg.h>
+#include <string.h>
+#ifdef	HAVE_STRINGS_H
+#include <strings.h>
+#endif
+#include <ctype.h>
+#include <time.h>
+#include <limits.h>
+
+#include "cobc.h"
+#include "tree.h"
+#include "replace.h"
+
+static struct cb_replace_list	*current_replace_list = NULL;
+static struct cb_replace_list	*save_current_replace = NULL;
+static struct cb_replace_list	*base_replace_list = NULL;
+
+struct cb_token_list {
+	struct cb_token_list	*next;			/* next pointer */
+	struct cb_token_list	*last;
+	const char		*text;
+	const char		*token;
+};
+
+static struct cb_token_list	*text_queue = NULL;
+static size_t			check_partial_match = 0;
+
+static int      ppecho_replace          (struct cb_replace_list *);
+
+static struct cb_token_list *
+pp_token_list_add (struct cb_token_list *list,
+		      const char *text,
+		      const char *token)
+{
+	struct cb_token_list	*p;
+	void			*tp;
+	int text_size = strlen (text);
```

Comment:
the suggested use of `cobc_plex_strdup` mentioned in the review for the previous commit has the benefit that we don't have the length fields at all (which otherwise should be changed to its real type `size_t`)